src/isolate/server/server.py

```diff
@@ -39,35 +39,51 @@ def Run(
         context: ServicerContext,
     ) -> Iterator[definitions.PartialRunResult]:
         messages: Queue[definitions.PartialRunResult] = Queue()
-        try:
-            environment = from_grpc(request.environment)
-        except ValueError:
-            return self.abort_with_msg(
-                f"Unknown environment kind: {request.environment.kind}.",
-                context,
-            )
-        except TypeError as exc:
-            return self.abort_with_msg(
-                f"Invalid environment parameter: {str(exc)}.",
-                context,
-            )
+        environments = []
+        for env in request.environments:
+            try:
+                environments.append(from_grpc(env))
+            except ValueError:
+                return self.abort_with_msg(
+                    f"Unknown environment kind",
+                    context,
+                )
+            except TypeError as exc:
+                return self.abort_with_msg(
+                    f"Invalid environment parameter: {str(exc)}.",
+                    context,
+                )
+
+        assert len(environments), "No environment definition found."
 
         run_settings = replace(
             self.default_settings,
             log_hook=partial(_add_log_to_queue, messages),
             serialization_method=request.function.method,
         )
-        environment.apply_settings(run_settings)
+
+        for environment in environments:
+            environment.apply_settings(run_settings)
+
+        environment = environments.pop(0)
```

Comment:
NIT: looks a bit more clear (instead of mutation)
```suggestion
        base_environment, *extra_environments = environments
```